content/blog/new-graphic-features/index.Rmd

```diff
@@ -0,0 +1,301 @@
+grid.points(
+  x = runif(5000), 
+  y = runif(5000), 
+  default.units = 'npc',
+  vp = viewport(clip = clip_path)
+)
+```
+
+Clipping is not only possible with single grobs. By combining grobs in a gList you can making the clipping region arbitrary complex:
+
+```{r}
+circle <- circleGrob(x = c(0.2, 0.8), r = 0.3)
+rect <- rectGrob(
+  width = unit(0.7, 'snpc'), 
+  height = unit(0.7, 'snpc'), 
+  vp = viewport(angle = 45)
+)
+clip_path <- gTree(children = gList(circle, rect))
+grid.points(
+  x = runif(5000), 
+  y = runif(5000), 
+  default.units = 'npc',
+  vp = viewport(clip = clip_path)
+)
+```
+
+The examples above seems quite contrived and decoupled from data visualization, but there are of course real world usages, e.g. clipping a 2D density estimate to the shape of a country or clipping data points inside a circular canvas for polar plots.
+
+The user interface for clipping paths are easy enough to understand, but it should be noted that there may be slight differences between devices as to which grob types can be used. Most notably, the use of text grobs for defining clipping paths is not something that will work for every device (but do work in ragg).
+
+### Alpha masks
+
+The last feature added to the graphics engine in this round is the ability of viewports to have an alpha mask assigned. When a mask is present, the grobs being drawn will apply the opacity of the mask. Note that this is different than a luminosity mask that uses the lightness of the mask as the alpha value. A mask can be any grob you want, or a collection of multiples:
+
+```{r}
+circle <- circleGrob(
+  x = c(0.2, 0.8), 
+  r = 0.3
+)
+rect <- rectGrob(
+  width = unit(0.7, 'snpc'), 
+  height = unit(0.7, 'snpc'), 
+  vp = viewport(angle = 45)
+)
+mask <- gTree(children = gList(circle, rect), gp = gpar(fill = "#00000066", col = NA))
+grid.rect(
+  x = c(0.25, 0.25, 0.75, 0.75), 
+  y = c(0.25, 0.75, 0.75, 0.25), 
+  width = 0.5,
+  height = 0.5,
+  gp = gpar(fill = c("steelblue", "firebrick")),
+  vp = viewport(mask = mask)
+)
+```
+
+As we see above, the areas in the mask where nothing is drawn have an opacity of 0, meaning that whatever is being drawn by the rectangle grob in these areas will be invisible. We also see that opacity is compounded by overlaying shapes as the areas covered both by the circle and the square in the mask has a higher opacity.
+
+## Future features
+
+It is of course dangerous to promise what the future brings. However, I do know of a few features adjacent to what have been discussed above that might make sense to know about.
+
+When it comes to clipping paths there is currently a lack of way to describe how multiple shapes are combined, since the fill rule is implicitly "winding" and you have no control over the direction the graphic device trace circles and rectangles. Work is already being done to let you control this from grid so it will become easier to e.g. punch out holes in a clipping path by overlaying two grobs.
```

Comment:
```suggestion
When it comes to clipping paths, there is currently a lack of way to describe how multiple shapes are combined, since the fill rule is implicitly "winding" and you have no control over the direction the graphic device trace circles and rectangles. Work is already being done to let you control this from grid, so it will become easier to e.g. punch out holes in a clipping path by overlaying two grobs.
```